### PR TITLE
added trace methon for AMP FP16

### PR DESCRIPTION
**Pull request by TezRomacH**
## Description

- Fixed Gradient Accumulation for FP16 (https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations)
- Added tracing support for FP16
- Added `trace` to SupervisedRunner
- Fixed segmentation tutorial download data bug
- Updated segmentation tutorial with `trace`

## Type of Change

- [x] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [ ] I have checked the code-style using `make check-style`.
- [ ] I have written the docstring in Google format for all the methods and classes that I used.
- [ ] I have checked the docs using `make check-docs`.

### CODE REVIEW

File: catalyst/dl/scripts/trace.py [hunk 1]
```diff
@@ -35,23 +58,40 @@ def trace_model_from_checkpoint(
...
-    model = experiment.get_model(next(iter(experiment.stages)))
+    if stage is None:
+        stage = list(experiment.stages)[0]
+
+    model = experiment.get_model(stage)
     checkpoint = utils.load_checkpoint(checkpoint_path)
     utils.unpack_checkpoint(checkpoint, model=model)
 
+    runner: RunnerType = RunnerType()
+    runner.set_model_device(model, device)
```
Comment (Scitator) [inline catalyst/dl/scripts/trace.py]: do we really need `set/get_model_device`? maybe, it's easier to just use
```
model, device = runner.model, runner.device
runner.model, runner.device = model, device
```
?

File: catalyst/dl/scripts/trace.py [hunk 2]
```diff
@@ -35,23 +58,40 @@ def trace_model_from_checkpoint(
...
+
+    if loader is None:
+        loader = 0
+    batch = experiment.get_native_batch(stage, loader)
+
     print("Tracing")
     traced = trace_model(
-        model, experiment, RunnerType,
+        model,
+        runner,
```
Comment (Scitator) [inline catalyst/dl/scripts/trace.py:78]: do we need to pass `runner` for tracing? 
is it not enough to just pass `model` and `runner.predict_batch`? 
we can get device with `utils.get_device()`

File: catalyst/dl/utils/torch.py [hunk 3]
```diff
@@ -1,34 +1,43 @@
...
+            for distributed and FP16 methond
+        device (Device, optional): device
+    """
     distributed_params = distributed_params or {}
     distributed_params = copy.deepcopy(distributed_params)
     if device is None:
         device = utils.get_device()
 
-    model = maybe_recursive_call(model, "to", device=device)
+    model: Model = maybe_recursive_call(model, "to", device=device)
```
Comment (Scitator) [inline catalyst/dl/utils/torch.py:40]: btw, I think we can remove `maybe_recursive_call` if we would use nn.ModuleDict as PyTorch model

File: catalyst/dl/core/experiment.py [hunk 4]
```diff
@@ -19,74 +19,115 @@ class Experiment(ABC):
...
+            _loader = loaders[loader]
+        elif isinstance(loader, int):
+            _loader = list(loaders.values())[loader]
+        else:
+            raise TypeError("Loader parameter must be a string or an integer")
+
+        dataset = _loader.dataset
+        collate_fn = _loader.collate_fn
+
+        sample = collate_fn([dataset[0]])
```
Comment (Scitator) [inline catalyst/dl/core/experiment.py]: if we add such method, what do you think about 
```
def get_native_batch(self, stage: str, loader: Union[str, int] = 0, batch_index: int):
sample = collate_fn([dataset[batch_index]])
```

by this way we can easily check the data samples from the Experiment

File: catalyst/dl/scripts/trace.py (hunk 2 above)
Comment (Scitator) [reply catalyst/dl/scripts/trace.py:78]: by this way we also will make our tracing support more general without any Experiment/Runner dependencies

File: examples/_tests_mnist_stages/__init__.py [hunk 5]
```diff
@@ -1,4 +1,9 @@
 # flake8: noqa
 from catalyst.dl import SupervisedRunner as Runner
 from .experiment import Experiment
-from .model import SimpleNet
+
+from catalyst.dl.registry import MODELS  # isort:skip
+
+#  TODO: fix hack for duplicate SimpleNets
+del MODELS["SimpleNet"]  # isort:skip
```
Comment (Scitator) [inline examples/_tests_mnist_stages/__init__.py]: I think you can remove this trick :)

File: examples/_tests_mnist_stages2/__init__.py [hunk 6]
```diff
@@ -1,6 +1,9 @@
 # flake8: noqa
 from catalyst.dl import registry, SupervisedRunner as Runner
 from .experiment import Experiment
-from .model import SimpleNet
 
-registry.Model(SimpleNet)
+from catalyst.dl.registry import MODELS  # isort:skip
+
+#  TODO: fix hack for duplicate SimpleNet
+del MODELS["SimpleNet"]  # isort:skip
```
Comment (Scitator) [inline examples/_tests_mnist_stages2/__init__.py]: and this

File: examples/cifar_simple/__init__.py [hunk 7]
```diff
@@ -1,4 +1,9 @@
 # flake8: noqa
 from catalyst.dl import SupervisedRunner as Runner
 from .experiment import Experiment
-from .model import SimpleNet
+
+from catalyst.dl.registry import MODELS  # isort:skip
+
+#  TODO: fix hack for duplicate SimpleNet
+del MODELS["SimpleNet"]  # isort:skip
```
Comment (Scitator) [inline examples/cifar_simple/__init__.py]: and this

File: examples/cifar_stages/__init__.py [hunk 8]
```diff
@@ -1,4 +1,9 @@
 # flake8: noqa
 from catalyst.dl import SupervisedRunner as Runner
 from .experiment import Experiment
-from .model import SimpleNet
+
+from catalyst.dl.registry import MODELS  # isort:skip
+
+#  TODO: fix hack for duplicate SimpleNet
+del MODELS["SimpleNet"]  # isort:skip
```
Comment (Scitator) [inline examples/cifar_stages/__init__.py]: and this

File: catalyst/dl/scripts/trace.py (hunk 1 above)
Comment (TezRomacH) [reply catalyst/dl/scripts/trace.py]: Yeah, i really think it's better to have this func either than copy-past `runner.model, runner.device`

File: catalyst/dl/scripts/trace.py (hunk 2 above)
Comment (TezRomacH) [reply catalyst/dl/scripts/trace.py:78]: Yes, we do need a runner, because we trace not a Tensor, but catalyst dict with features and whatever.

> we can get device with utils.get_device()

no, because Runner may have `cuda` and we need to trace in `cpu`, or runner may have cpu, but fp16 needs `cuda`, etc


> by this way we also will make our tracing support more general without any Experiment/Runner dependencies

not agreed, because you may trace your model easily with torch.jit, but Runner is needed to make forward pass with dict with additional `any2device`

File: catalyst/dl/core/experiment.py (hunk 4 above)
Comment (TezRomacH) [reply catalyst/dl/core/experiment.py]: ok

File: examples/_tests_mnist_stages/__init__.py (hunk 5 above)
Comment (TezRomacH) [reply examples/_tests_mnist_stages/__init__.py]: is the registry already fixed?

File: catalyst/dl/scripts/trace.py (hunk 1 above)
Comment (Scitator) [reply catalyst/dl/scripts/trace.py]: why? it looks like additional codebase

File: catalyst/dl/scripts/trace.py (hunk 2 above)
Comment (Scitator) [reply catalyst/dl/scripts/trace.py:78]: so, do we need `runner` or we need just `runner.predict_batch`?
and do we need to carry about `runner.device` if we need to trace the model for prespecified one?

File: examples/_tests_mnist_stages/__init__.py (hunk 5 above)
Comment (Scitator) [reply examples/_tests_mnist_stages/__init__.py]: yes

Comment (Scitator) [reply examples/_tests_mnist_stages/__init__.py]: it was pytest issue

File: catalyst/dl/scripts/trace.py (hunk 1 above)
Comment (TezRomacH) [reply catalyst/dl/scripts/trace.py]: because it's easy to read and less probability to make a mistake :)

File: catalyst/dl/scripts/trace.py (hunk 2 above)
Comment (TezRomacH) [reply catalyst/dl/scripts/trace.py:78]: Yes, we need predictions through the runner. Because you may implement it in other ways :)

File: catalyst/dl/utils/torch.py (hunk 3 above)
Comment (TezRomacH) [reply catalyst/dl/utils/torch.py:40]: I would like to leave it for this PR, and check in the another

File: examples/_tests_mnist_stages/__init__.py [hunk 9]
```diff
@@ -1,4 +1,6 @@
 # flake8: noqa
-from catalyst.dl import SupervisedRunner as Runner
+from catalyst.dl import registry, SupervisedRunner as Runner
 from .experiment import Experiment
 from .model import SimpleNet
+
+registry.Model(SimpleNet)
```
Comment (Scitator) [inline examples/_tests_mnist_stages/__init__.py]: why do we need this if we already have registred SimpleNet?
https://github.com/catalyst-team/catalyst/blob/master/examples/_tests_mnist_stages/model.py#L7

Comment (Scitator) [inline examples/_tests_mnist_stages/__init__.py]: why do wee need this, if we have already registred `SimpleNet`?
https://github.com/catalyst-team/catalyst/blob/master/examples/_tests_mnist_stages/model.py#L7
